### PR TITLE
V8: Don't allow logins for users with no content and/or media start nodes

### PR DESCRIPTION
**Pull request by kjac**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes https://github.com/umbraco/Umbraco-CMS/issues/5641

### Description

This PR ensures that users that belong to groups that do not define any content and/or media start nodes are not allowed to log in to the back office.

If a login for such a user is attempted, the login will fail and a message will be added to the log along the lines of: `Login attempt failed for username [username] from IP address [user IP], no content and/or media start nodes could be found for any of the user's groups`.

image

/cc @Shazwazza 

#### Testing this PR

1. Create a new group that has no content and media start nodes configured.
2. Create a new user and make it a member of that group (and no other groups).
3. Verify that the user can't log in to the back office.
4. Assign any other group to the user as well (that is, any group _with_ content and media start nodes configured).
5. Verify that the user can log in.

### CODE REVIEW

**Comment by emmaburstow** [comment]
Hiiiii Kennnnn,

Thanks for the work here. We'll let you know if we have any questions.

Em

File: src/Umbraco.Web/Security/BackOfficeSignInManager.cs [hunk 1]
```diff
@@ -112,6 +112,19 @@ private async Task<SignInStatus> PasswordSignInAsyncImpl(string userName, string
                     return SignInStatus.LockedOut;
                 }
 
+                // We need to verify that the user belongs to one or more groups that define content and media start nodes.
+                // To do so we have to create the user claims identity and validate the calculated start nodes.
+                var userIdentity = await CreateUserIdentityAsync(user);
+                if(userIdentity is UmbracoBackOfficeIdentity backOfficeIdentity)
```
Comment (ronaldbarendse) [inline src/Umbraco.Web/Security/BackOfficeSignInManager.cs]:
```suggestion
                if (userIdentity is UmbracoBackOfficeIdentity backOfficeIdentity)
```

File: src/Umbraco.Web/Security/BackOfficeSignInManager.cs [hunk 2]
```diff
@@ -112,6 +112,19 @@ private async Task<SignInStatus> PasswordSignInAsyncImpl(string userName, string
...
+                {
+                    if(backOfficeIdentity.StartContentNodes.Length == 0 || backOfficeIdentity.StartMediaNodes.Length == 0)
```
Comment (ronaldbarendse) [inline src/Umbraco.Web/Security/BackOfficeSignInManager.cs]:
```suggestion
                    if (backOfficeIdentity.StartContentNodes.Length == 0 || backOfficeIdentity.StartMediaNodes.Length == 0)
```

**Comment by kjac** [comment]
PR updated with latest v8/dev

**Pull request merged**